### PR TITLE
Feat/blockhash

### PR DESCRIPTION
**Pull request by Eikix**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves: #149 

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: crates/evm/src/instructions/block_information.cairo [hunk 1]
```diff
@@ -15,7 +16,27 @@ impl BlockInformation of BlockInformationTrait {
...
+        // If input block number is higher than current_block - 10, return 0
+        // Note: in the specs, input block number can be equal - at most - to the current block number minus one.
+        // In Starknet, the `get_block_hash_syscall` is capped at current block minus ten.
+        // TODO: monitor the changes in the `get_block_hash_syscall` syscall.
+        // source: https://docs.starknet.io/documentation/architecture_and_concepts/Smart_Contracts/system-calls-cairo1/#get_block_hash
+        if block_number + 10 > current_block || block_number + 256 < current_block {
+            return self.stack.push(0);
+        }
+
+        let block_hash = get_block_hash_syscall(block_number);
```
Comment (enitrat) [inline crates/evm/src/instructions/block_information.cairo:32]:
```suggestion
        let maybe_block_hash = get_block_hash_syscall(block_number);
```

Comment (enitrat) [reply crates/evm/src/instructions/block_information.cairo:32]: proposing `maybe` prefixes for Result<T> types

File: crates/evm/src/stack.cairo [hunk 2]
```diff
@@ -113,11 +114,25 @@ impl StackImpl of StackTrait {
     #[inline(always)]
     fn pop_usize(ref self: Stack) -> Result<usize, EVMError> {
         let item: u256 = self.pop()?;
-        // item.try_into().ok_or(EVMError::TypeConversionError(TYPE_CONVERSION_ERROR))
         let item: usize = item.try_into_result()?;
         Result::Ok(item)
     }
 
+    /// Calls `Stack::pop` and tries to convert it to usize
```
Comment (enitrat) [inline crates/evm/src/stack.cairo:121]:
```suggestion
    /// Calls `Stack::pop` and tries to convert it to u64
```

File: crates/evm/src/tests/test_instructions/test_block_information.cairo [hunk 3]
```diff
@@ -9,6 +9,58 @@ use openzeppelin::token::erc20::interface::IERC20CamelDispatcherTrait;
...
+    let mut machine = setup_machine();
+
+    set_block_number(500);
+
+    // When
+    machine.stack.push(491).unwrap();
+    machine.exec_blockhash();
+
+    // Then
+    assert(machine.stack.peek().unwrap() == 0, 'stack top should be 1692873993');
```
Comment (enitrat) [inline crates/evm/src/tests/test_instructions/test_block_information.cairo:42]: error string not matching expected result

File: crates/evm/src/tests/test_instructions/test_block_information.cairo [hunk 4]
```diff
@@ -9,6 +9,58 @@ use openzeppelin::token::erc20::interface::IERC20CamelDispatcherTrait;
 use starknet::testing::{set_block_timestamp, set_block_number, set_contract_address};
 use utils::constants::CHAIN_ID;
 
+/// 0x40 - BLOCKHASH
+#[test]
+#[available_gas(20000000)]
+fn test_block_hash_below_bounds() {
```
Comment (enitrat) [inline crates/evm/src/tests/test_instructions/test_block_information.cairo:15]:
```suggestion
fn test_exec_blockhash_below_bounds() {
```

File: crates/evm/src/tests/test_instructions/test_block_information.cairo [hunk 5]
```diff
@@ -9,6 +9,58 @@ use openzeppelin::token::erc20::interface::IERC20CamelDispatcherTrait;
...
+    machine.stack.push(243).unwrap();
+    machine.exec_blockhash();
+
+    // Then
+    assert(machine.stack.peek().unwrap() == 0, 'stack top should be 1692873993');
+}
+
+#[test]
+#[available_gas(20000000)]
+fn test_block_hash_above_bounds() {
```
Comment (enitrat) [inline crates/evm/src/tests/test_instructions/test_block_information.cairo:31]:
```suggestion
fn test_exec_blockhash_above_bounds() {
```

File: crates/evm/src/tests/test_instructions/test_block_information.cairo [hunk 6]
```diff
@@ -9,6 +9,58 @@ use openzeppelin::token::erc20::interface::IERC20CamelDispatcherTrait;
...
+    // Then
+    assert(machine.stack.peek().unwrap() == 0, 'stack top should be 1692873993');
+}
+
+// TODO: implement exec_blockhash testing for block number within bounds
+// https://github.com/starkware-libs/cairo/blob/77a7e7bc36aa1c317bb8dd5f6f7a7e6eef0ab4f3/crates/cairo-lang-starknet/cairo_level_tests/interoperability.cairo#L173
+#[ignore]
+#[test]
+#[available_gas(20000000)]
+fn test_block_hash_within_bounds() {
```
Comment (enitrat) [inline crates/evm/src/tests/test_instructions/test_block_information.cairo:50]:
```suggestion
fn test_exec_blockhash_within_bounds() {
```

File: crates/utils/src/traits.cairo [hunk 7]
```diff
@@ -85,11 +85,11 @@ trait TryIntoResult<T, U> {
     fn try_into_result(self: T) -> Result<U, EVMError>;
 }
 
-impl U256TryIntoResultU32 of TryIntoResult<u256, usize> {
+impl U256TryIntoResultU32<U, +TryInto<u256, U>> of TryIntoResult<u256, U> {
```
Comment (enitrat) [inline crates/utils/src/traits.cairo:88]: Nice! But if you made this impl generic over U, we need to change the impl name and codedoc

**Comment by Eikix** [comment]
Thanks for the review. Too many things were overlooked on my side, will keep an eye out 🎯

**Review by enitrat — APPROVED** [review]

**Review by enitrat — APPROVED** [review]

**Pull request merged**